File: packages/septr-go/bola_test.go
package septr

import "testing"

func TestExtractRouteParams_Express(t *testing.T) {
	p := extractRouteParams("/api/users/:userId/orders/:orderId")
	if len(p) < 2 || p[0] != "userId" || p[1] != "orderId" {
		t.Fatalf("expected userId,orderId got %v", p)
	}
}

func TestExtractRouteParams_Static(t *testing.T) {
	if len(extractRouteParams("/api/health")) != 0 {
		t.Fatal("expected empty")
	}
}

func TestTokenClaims_Valid(t *testing.T) {
	c := extractTokenClaims("eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwidXNlcl9pZCI6IjU2NyIsImlkIjoiODkwIn0.signature")
	if c["sub"] != "1234567890" || c["user_id"] != "567" || c["id"] != "890" {
		t.Fatalf("unexpected claims: %v", c)
	}
}

func TestTokenClaims_Invalid(t *testing.T) {
	if len(extractTokenClaims("not-a-token")) != 0 {
		t.Fatal("expected empty")
	}
}

func TestDetectBOLA_RouteParams(t *testing.T) {
	ev := detectBOLA([]string{"userId"}, nil, map[string]string{"sub": "user_123"}, "/api/users/:userId", "GET", nil)
	if ev == nil || ev.PatternID != "bola_param_mismatch" {
		t.Fatal("expected bola_param_mismatch")
	}
}

func TestDetectBOLA_BodyFields(t *testing.T) {
	ev := detectBOLA(nil, map[string]string{"userId": "user_456"}, map[string]string{"sub": "user_123"}, "/api/users", "POST", nil)
	if ev == nil || ev.PatternID != "bola_body_mismatch" {
		t.Fatal("expected bola_body_mismatch")
	}
}

func TestDetectBOLA_NoToken(t *testing.T) {
	ev := detectBOLA([]string{"userId"}, nil, map[string]string{}, "/api/users/:userId", "GET", nil)
	if ev != nil {
		t.Fatal("expected nil when no token")
	}
}

func TestDetectBOLA_NonIDParam(t *testing.T) {
	ev := detectBOLA([]string{"99"}, nil, map[string]string{"sub": "42"}, "/api/users/:userId", "GET", nil)
	if ev != nil {
		t.Fatal("expected nil for non-ID param")
	}
}

func TestMatchRouteTemplate(t *testing.T) {
	if got := MatchRouteTemplate("/api/users/999", []string{"/api/users/:userId"}); got != "/api/users/:userId" {
		t.Fatalf("expected template match, got %q", got)
	}
	if got := MatchRouteTemplate("/api/users/999", []string{"/api/users/{user_id}"}); got != "/api/users/{user_id}" {
		t.Fatalf("expected braces template match, got %q", got)
	}
	if got := MatchRouteTemplate("/api/users/999/orders", []string{"/api/users/:userId"}); got != "" {
		t.Fatalf("expected no match, got %q", got)
	}
	if got := MatchRouteTemplate("/api/health", []string{"/api/users/:userId"}); got != "" {
		t.Fatalf("expected no match for static path, got %q", got)
	}
}

func TestExtractRouteParamValues(t *testing.T) {
	values := ExtractRouteParamValues("/api/users/:userId", "/api/users/999")
	if values["userId"] != "999" {
		t.Fatalf("expected userId=999, got %v", values)
	}
	if len(ExtractRouteParamValues("/api/users/:userId", "/api/users")) != 0 {
		t.Fatal("expected empty for mismatched lengths")
	}
}

func TestDetectBOLA_ParamValueMismatch(t *testing.T) {
	ev := detectBOLA([]string{"userId"}, nil, map[string]string{"sub": "42"}, "/api/users/:userId", "GET", map[string]string{"userId": "999"})
	if ev == nil || ev.PatternID != "bola_param_mismatch" {
		t.Fatal("expected bola_param_mismatch on value mismatch")
	}
}

func TestDetectBOLA_ParamValueMatch(t *testing.T) {
	ev := detectBOLA([]string{"userId"}, nil, map[string]string{"sub": "42"}, "/api/users/:userId", "GET", map[string]string{"userId": "42"})
	if ev != nil {
		t.Fatal("expected nil when param value matches token")
	}
}

func TestDetectBOLA_NonIDParamValue(t *testing.T) {
	ev := detectBOLA([]string{"category"}, nil, map[string]string{"sub": "42"}, "/api/items/:category", "GET", map[string]string{"category": "books"})
	if ev != nil {
		t.Fatal("expected nil for non-ID param values")
	}
}

File: packages/septr-node/src/__tests__/bola.test.ts
import { describe, it, expect } from "vitest"
import { extractRouteParams, extractTokenClaims, extractRouteParamValues, matchRouteTemplate, detectBOLA } from "../core/bola"

describe("extractRouteParams", () => {
  it("extracts Express-style params", () => {
    const params = extractRouteParams("/api/users/:userId/orders/:orderId")
    expect(params).toEqual(["userId", "orderId"])
  })

  it("extracts Hono-style params", () => {
    const params = extractRouteParams("/api/users/:userId")
    expect(params).toEqual(["userId"])
  })

  it("returns empty for static routes", () => {
    const params = extractRouteParams("/api/health")
    expect(params).toEqual([])
  })

  it("handles empty path", () => {
    expect(extractRouteParams("")).toEqual([])
  })
})

describe("extractTokenClaims", () => {
  it("extracts claims from valid JWT", () => {
    const token = "eyJhbGciOiJIUzI1NiJ9." + "eyJzdWIiOiIxMjM0NTY3ODkwIiwidXNlcl9pZCI6IjU2NyIsImlkIjoiODkwIn0.signature"
    const claims = extractTokenClaims(token)
    expect(claims.sub).toBe("1234567890")
    expect(claims.user_id).toBe("567")
    expect(claims.id).toBe("890")
  })

  it("returns empty for invalid token", () => {
    const claims = extractTokenClaims("not-a-token")
    expect(claims).toEqual({})
  })

  it("returns empty for token with wrong parts", () => {
    const claims = extractTokenClaims("part1.part2")
    expect(claims).toEqual({})
  })

  it("handles empty token", () => {
    expect(extractTokenClaims("")).toEqual({})
  })
})

describe("detectBOLA", () => {
  it("detects BOLA from route params", () => {
    const result = detectBOLA(
      ["userId"],
      null,
      { sub: "user_123" },
      "/api/users/:userId",
      "GET",
    )
    expect(result).not.toBeNull()
    expect(result!.type).toBe("bola")
    expect(result!.patternId).toBe("bola_param_mismatch")
  })

  it("detects BOLA from body fields", () => {
    const result = detectBOLA(
      [],
      { userId: "user_456" },
      { sub: "user_123" },
      "/api/users",
      "POST",
    )
    expect(result).not.toBeNull()
    expect(result!.type).toBe("bola")
    expect(result!.patternId).toBe("bola_body_mismatch")
  })

  it("returns null when token has no user claim", () => {
    const result = detectBOLA(
      ["userId"],
      null,
      {},
      "/api/users/:userId",
      "GET",
    )
    expect(result).toBeNull()
  })

  it("returns null when no route params and no body", () => {
    const result = detectBOLA(
      [],
      null,
      { sub: "user_123" },
      "/api/health",
      "GET",
    )
    expect(result).toBeNull()
  })

  it("does not flag non-ID route params", () => {
    const result = detectBOLA(
      ["category"],
      null,
      { sub: "user_123" },
      "/api/items/:category",
      "GET",
    )
    expect(result).toBeNull()
  })

  it("passes through when body ID matches token", () => {
    const result = detectBOLA(
      [],
      { userId: "user_123" },
      { sub: "user_123" },
      "/api/users",
      "PUT",
    )
    expect(result).toBeNull()
  })

  it("sets severity to critical for body mismatch", () => {
    const result = detectBOLA(
      [],
      { userId: "user_456" },
      { sub: "user_123" },
      "/api/users",
      "POST",
    )
    expect(result!.severity).toBe("critical")
  })
})

describe("matchRouteTemplate", () => {
  it("matches concrete path to template", () => {
    expect(matchRouteTemplate("/api/users/999", ["/api/users/:userId"])).toBe("/api/users/:userId")
    expect(matchRouteTemplate("/api/users/999", ["/api/users/{user_id}"])).toBe("/api/users/{user_id}")
  })

  it("returns null when no structural match", () => {
    expect(matchRouteTemplate("/api/users/999/orders", ["/api/users/:userId"])).toBeNull()
    expect(matchRouteTemplate("/api/health", ["/api/users/:userId"])).toBeNull()
  })
})

describe("extractRouteParamValues", () => {
  it("extracts param values from concrete path", () => {
    expect(extractRouteParamValues("/api/users/:userId", "/api/users/999")).toEqual({ userId: "999" })
  })

  it("returns empty for mismatched lengths", () => {
    expect(extractRouteParamValues("/api/users/:userId", "/api/users")).toEqual({})
  })
})

describe("detectBOLA with route param values", () => {
  it("flags param value that does not match the authenticated user", () => {
    const result = detectBOLA(
      ["userId"],
      null,
      { sub: "42" },
      "/api/users/:userId",
      "GET",
      { userId: "999" },
    )
    expect(result).not.toBeNull()
    expect(result!.patternId).toBe("bola_param_mismatch")
  })

  it("passes through when param value matches the authenticated user", () => {
    const result = detectBOLA(
      ["userId"],
      null,
      { sub: "42" },
      "/api/users/:userId",
      "GET",
      { userId: "42" },
    )
    expect(result).toBeNull()
  })

  it("ignores values of non-ID params", () => {
    const result = detectBOLA(
      ["category"],
      null,
      { sub: "42" },
      "/api/items/:category",
      "GET",
      { category: "books" },
    )
    expect(result).toBeNull()
  })
})
